### PR TITLE
Queue | Edit issue diagnostic code

### LINKED ISSUE #5202: Users can't edit the diagnostic code of an issue
### View issues and decide that you want to edit issue 5
image
### But the dropdowns don't allow editing of diagnostic codes
image

# AC
- The Edit Issue screen should allow you to edit the diagnostic codes

### PR DESCRIPTION
**Pull request by dannysellers**
Resolves #5202 

### Description
Within the `AddEditIssueView`, we previously didn't allow users to set diagnostic codes (i.e. `8312: Hypoglossal nerve neuritis`)

### Acceptance Criteria 
- [x] Confirm edits made to issue diagnostic codes persist after saving
- [x] Diagnostic code dropdown should only be enabled in cases described by [Appendix F - VACOLS Issue Coding.pdf](https://git.io/vxBtN)

### Testing Plan
1. Go to Draft Decision checkout flow, add or edit issue, set diagnostic code
2. Save edits, confirm code is updated in SelectDispositionsView, AppealDetail

### CODE REVIEW

File: constants/DIAGNOSTIC_CODE_DESCRIPTIONS.json [hunk 1]
```diff
@@ -0,0 +1,805 @@
+{
```
[the remark this answers is not in the document]
Comment (dannysellers) [reply constants/DIAGNOSTIC_CODE_DESCRIPTIONS.json]: @amprokop It doesn't look like these are stored in `ISSREF`—the [dump](https://github.com/department-of-veterans-affairs/caseflow/blob/master/local/vacols/VACOLS::Issref_dump.csv) just has `## Diagnostic Code`, same as the PDF. It looks like these might come from [`VFTYPES.ftdesc`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/local/vacols/VACOLS::Vftypes_dump.csv)—are those the VACOLS descriptions we should be using @cmgiven?

[the remark this answers is not in the document]
Comment (dannysellers) [reply constants/DIAGNOSTIC_CODE_DESCRIPTIONS.json]: Using `VFTYPES` does give us more specific info. Thoughts?

screen shot 2018-04-24 at 3 57 05 pm

[the remark this answers is not in the document]
Comment (dannysellers) [reply constants/DIAGNOSTIC_CODE_DESCRIPTIONS.json]: TODO: As Chris and Anya and I discussed, we'll ultimately want to fetch and cache these values from VACOLS. In the meantime, though, I created two sets of constants—one for Status users (vets) and the other for Board staff.

**Comment by amprokop** [comment]
image

image

It looks like we're preventing the editing of level 2 codes in the very common Compensation -> Service Connection -> All Others pathway — wondering if it's been discussed whether or not to just remove the level 2 dropdown entirely. 

This isn't something that should necessarily block merging this but wanted to bring it up — cc @laurjpeterson @sneha-pai

**Comment by amprokop** [comment]
Also, when we edit an issue's program/codes, we do not currently clear the `Note` field. 

My guess is that typically when an attorney wants to *edit* an issue rather than deleting it and adding a new one, that issue would typically be somewhat adjacent/related to the current issue, so maybe keeping the note field is better UX — also, if the note went away without warning, potentially that's jarring. But raising this with @laurjpeterson and @sneha-pai

**Comment by amprokop** [comment]
`Compensation -> Increased rating -> Schedular & Extraschedular` should contain diagnostic code, while  `Compensation -> Increased rating -> Other should not`
image

But if I have an issue with `Schedular & Extraschedular`:
image
And I edit it to `Other`:
image

**Comment by amprokop** [comment]
@dannysellers, I think maybe there could be two bugs here:

1. Diagnostic code should be wiped in the UI if we change to an issue type without it
2. If an issue is invalid, we should ensure an error is passed to the frontend and the change is not made

cc @aroltsch

**Comment by amprokop** [comment]
image

**Comment by dannysellers** [comment]
I believe the issue with the diagnostic code persisting is due to the `descriptions` attribute, which comes from the server and is `${code} - ${description}` per issue level. We previously were using statements like `getDiagnosticCodeLabel(_.last(codes)) || _.last(description)` to display the diagnostic code info, but we should only display a `code` for diagnostic code if applicable.

**Comment by dannysellers** [comment]
A thought: The usage of [`saveEditedAppealIssue` in `AddEditIssueView`](https://github.com/department-of-veterans-affairs/caseflow/blob/27ebe1aca57bec247a1cad336df001bd5a7bbf36/client/app/queue/AddEditIssueView.jsx#L132) is beginning to feel like the odd one out. Other places it's used (in [`SelectDispositionsView`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/queue/SelectDispositionsView.jsx#L71) and [`IssueRemandReasonsOptions`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/queue/components/IssueRemandReasonsOptions.jsx#L58)), we only update `editingIssue` as part of a larger flow, while here, the update to `editingIssue` is almost incidental—immediately after updating it, we update the appeal issues in `stagedChanges` and in `loadedQueue` with more complete information (what comes back from the server).

It might make sense to bypass updating `editingIssue` here, and just `editStagedAppeal` and `editAppeal`. We have to do both: updates to the staged appeal will be rendered in `SelectDispositionsView` (the page before/after `AddEditIssueView`), and updating/creating issues will affect the issue count in the table views (which use `loadedQueue`)

**Pull request merged**